src/ndarray/autograd.cc

```diff
@@ -0,0 +1,120 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file autograd.cc
+ * \brief (TODO)
+ */
+
+#include <mxnet/operator.h>
+#include <mxnet/executor.h>
+#include <iostream>
+#include "./autograd.h"
+
+namespace mxnet {
+namespace ndarray {
+
+AutogradRuntime* AutogradRuntime::instance_ = new AutogradRuntime();
+
+nnvm::NodeEntry AttrHint(nnvm::NodeEntry src, nnvm::NodeEntry like) {
+  static const Op* id_like = Op::Get("_identity_with_attr_like_rhs");
+  nnvm::NodePtr n = nnvm::Node::Create();
+  n->attrs.op = id_like;
+  n->attrs.name = src.node->attrs.name + "_id";
+  n->inputs = {src, like};
+  return nnvm::NodeEntry{n, 0, 0};
+}
+
+void AutogradRuntime::StartRecording() {
+  is_recording_ = true;
+}
+
+void AutogradRuntime::StopRecording() {
+  is_recording_ = false;
+}
+
+bool AutogradRuntime::IsRecording() {
+  return is_recording_;
+}
+
+void AutogradRuntime::MarkForRecord(const std::vector<NDArray*>& arrays) {
+  for (const NDArray* arr : arrays) {
+    bp_flags[arr] = true;
+  }
+}
+
+void AutogradRuntime::UnMarkForRecord(const std::vector<NDArray*>& arrays) {
+  for (const NDArray* arr : arrays) {
+    bp_flags[arr] = false;
+  }
+}
+
+void AutogradRuntime::RecordOp(const nnvm::Op* op,
+                               const nnvm::NodeAttrs& attrs,
+                               std::vector<NDArray>& inputs,
+                               std::vector<NDArray>& outputs) {
+  nnvm::NodePtr node = nnvm::Node::Create();
+  // (TODO) name of operator
+  node->attrs = attrs;
+  node->attrs.name = op->name + "_temp_";
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    inputs[i].entry_->node->attrs.name = "input_" + std::to_string(i);
+    node->inputs.push_back(*(inputs[i].entry_));
+  }
+
+  for (size_t i = 0; i < outputs.size(); ++i) {
+    outputs[i].entry_->node = node;
+    outputs[i].entry_->index = i;
+    outputs[i].entry_->version++;
+  }
+}
+
+// (TODO) get backward graph by nnvm directly
+// (TODO) execute the backward graph
+std::vector<NDArray> AutogradRuntime::ComputeBackward(std::vector<NDArray>& inputs,
+                                      std::vector<NDArray>& outputs) {
+
+  nnvm::Symbol sym;
+  for (size_t i = 0; i < outputs.size(); ++i) {
+    sym.outputs.push_back(*(outputs[i].entry_));
+  }
+  sym.Print(std::cout);
+
+  Context ctx = Context::CPU(); // (TODO) fixme
+  std::map<std::string, Context> ctx_map;
+  std::vector<NDArray> aux_states;
+
+  std::vector<NDArray> grads;
+  std::vector<OpReqType> grad_reqs;
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    NDArray grad(inputs[i].shape(), inputs[i].ctx());
+    grad = static_cast<real_t>(1.0);
+    grads.push_back(grad);
+    grad_reqs.push_back(OpReqType::kWriteTo);
+  }
+
+  Executor *exec = Executor::Bind(sym,
```

Comment:
Nop, see [TODOs](https://github.com/dmlc/mxnet/pull/5129/files/5c24b3de4a517bb5e819e2267cce9ac47ac518d8#diff-377ece7abda57acc55d9c3a7892e61a7R70). Just raise the PR for MinPy meeting and tracking the progress. Many places will be rewriten.